File: .agents/skills/op-search/query.mjs
#!/usr/bin/env node
// Semantic / vector search over the Open Pulse extractor indices (RAG).
//
// The Git Metadata Extractor service (GME API v3) indexes every source into a
// per-provider vector store and serves semantic search with reranking — each
// hit carries a vector_score (embedding similarity) and rerank_score
// (cross-encoder). Distinct from op-collections (exact DuckDB lookups).
//
// Reads OPENPULSE_ENDPOINT (base host) and OPENPULSE_AUTH (user/password) from
// the nearest .env or process.env; appends /api/extractor. Read-only.
//
// Subcommands: search <provider> <query> | stats <provider> | manifest | freshness | get <path>
// Flags: --top-k --candidate-k --target --filter <json> --sources --param k=v --raw
//
// Usage:
//   node query.mjs search github_repos "protein structure prediction" --top-k 5
//   node query.mjs search orcid "machine learning EPFL" --target persons
//   node query.mjs manifest

import { readFile, stat } from 'node:fs/promises';
import { dirname, join } from 'node:path';
import { fileURLToPath } from 'node:url';

const EXTRACTOR_BASE = '/api/extractor';

async function loadDotenv() {
	for (let dir of [process.cwd(), dirname(fileURLToPath(import.meta.url))]) {
		for (let i = 0; i < 10; i++) {
			const envPath = join(dir, '.env');
			try {
				await stat(envPath);
				const text = await readFile(envPath, 'utf8');
				for (const line of text.split('\n')) {
					const t = line.trim();
					if (!t || t.startsWith('#') || !t.includes('=')) continue;
					const j = t.indexOf('=');
					const k = t.slice(0, j).trim();
					if (process.env[k] === undefined) process.env[k] = t.slice(j + 1).trim();
				}
				return;
			} catch {
				const parent = dirname(dir);
				if (parent === dir) break;
				dir = parent;
			}
		}
	}
}

function parseArgs(argv) {
	const pos = [];
	const o = { param: [] };
	for (let i = 0; i < argv.length; i++) {
		const a = argv[i];
		if (a === '--raw') o.raw = true;
		else if (a === '--sources') o.sources = true;
		else if (a === '--top-k') o.top_k = Number(argv[++i]);
		else if (a === '--candidate-k') o.candidate_k = Number(argv[++i]);
		else if (a === '--target') o.target = argv[++i];
		else if (a === '--filter') o.filter = argv[++i];
		else if (a === '--param') o.param.push(argv[++i]);
		else pos.push(a);
	}
	return { pos, o };
}

async function request(method, path, params, body, raw) {
	const endpoint = process.env.OPENPULSE_ENDPOINT;
	const auth = process.env.OPENPULSE_AUTH;
	if (!endpoint || !auth || !auth.includes('/')) {
		console.error('error: OPENPULSE_ENDPOINT and OPENPULSE_AUTH (user/password) must be set');
		process.exit(2);
	}
	const [user, ...rest] = auth.split('/');
	const password = rest.join('/');
	const qs = new URLSearchParams();
	for (const [k, v] of Object.entries(params || {})) if (v !== undefined && v !== null && v !== '') qs.set(k, v);
	let url = `${endpoint.replace(/\/$/, '')}${EXTRACTOR_BASE}${path}`;
	if ([...qs].length) url += `?${qs}`;
	const token = Buffer.from(`${user}:${password}`).toString('base64');
	const headers = { Authorization: `Basic ${token}`, Accept: 'application/json' };
	const init = { method, headers };
	if (body !== undefined) {
		headers['Content-Type'] = 'application/json';
		init.body = JSON.stringify(body);
	}
	let res;
	try {
		res = await fetch(url, init);
	} catch (e) {
		console.error(`network error: ${e.message}`);
		process.exit(1);
	}
	const text = await res.text();
	if (!res.ok) {
		console.error(`http ${res.status}: ${text}`);
		process.exit(1);
	}
	if (raw) {
		process.stdout.write(text.endsWith('\n') ? text : text + '\n');
		return;
	}
	try {
		process.stdout.write(JSON.stringify(JSON.parse(text), null, 2) + '\n');
	} catch {
		process.stdout.write(text.endsWith('\n') ? text : text + '\n');
	}
}

async function main() {
	await loadDotenv();
	const { pos, o } = parseArgs(process.argv.slice(2));
	const c = pos.shift();
	const a = pos.shift();
	const extra = {};
	for (const pair of o.param) {
		const i = pair.indexOf('=');
		extra[pair.slice(0, i).trim()] = pair.slice(i + 1).trim();
	}
	if (c === 'search') {
		const query = pos.join(' ').trim();
		if (!a || !query) {
			console.error('error: usage: search <provider> <query>');
			process.exit(2);
		}
		const body = { query, top_k: o.top_k, candidate_k: o.candidate_k, target: o.target };
		if (o.filter) body.filter_payload = JSON.parse(o.filter);
		for (const k of Object.keys(body)) if (body[k] === undefined) delete body[k];
		return request('POST', `/v2/indices/${a}/search`, extra, body, o.raw);
	}
	if (c === 'stats') return request('GET', `/v2/indices/${a}/stats`, extra, undefined, o.raw);
	if (c === 'manifest') return request('GET', '/v2/manifest', { sources: o.sources ? 'true' : undefined, ...extra }, undefined, o.raw);
	if (c === 'freshness') return request('GET', '/v2/indices/freshness', extra, undefined, o.raw);
	if (c === 'get') return request('GET', a.startsWith('/') ? a : `/${a}`, extra, undefined, o.raw);
	console.error(`error: unknown command '${c}'`);
	process.exit(2);
}

main();

File: .claude/skills/op-search/query.py
#!/usr/bin/env python3
"""Semantic / vector search over the Open Pulse extractor indices (RAG).

The Git Metadata Extractor service (GME API v3) indexes every source into a
per-provider vector store (Qdrant for most) and serves semantic search with
reranking: each hit carries a `vector_score` (embedding similarity) and a
`rerank_score` (cross-encoder). This is the RAG retrieval layer — distinct
from op-collections, which does exact keyword/structured lookups over the
hub's DuckDB mirror.

Reads OPENPULSE_ENDPOINT (base host, e.g. https://openpulse.epfl.ch) and
OPENPULSE_AUTH (user/password) from .env. The skill appends /api/extractor.
HTTP Basic, password-only. All subcommands here are read-only.

Subcommands:
    search <provider> <query>     semantic search (POST); returns ranked hits
    stats <provider>              doc/chunk counts + last_updated for one index
    manifest                      all providers: entity_types, backend, id_shape
    freshness                     per-provider counts + age since last update
    get <path>                    arbitrary GET under /api/extractor

Flags (search):
    --top-k N         max results (default 10)
    --candidate-k N   vector candidates fetched before reranking
    --target T        entity type for multi-entity indices (e.g. orcid -> persons)
    --filter JSON     metadata filter dict (index-specific; Qdrant shape for most)
Flags (general):
    --param k=v       extra query param (repeatable)
    --raw             print body verbatim

Usage:
    python query.py search github_repos "protein structure prediction" --top-k 5
    python query.py search orcid "machine learning EPFL" --target persons
    python query.py search openalex "graph neural networks" --target works --candidate-k 50
    python query.py stats zenodo_records
    python query.py manifest
"""
from __future__ import annotations

import argparse
import base64
import json
import os
import socket
import sys
import urllib.error
import urllib.parse
import urllib.request
from pathlib import Path

EXTRACTOR_BASE = "/api/extractor"


def load_dotenv() -> None:
    for start in (Path.cwd(), Path(__file__).resolve().parent):
        for parent in [start, *start.parents]:
            env = parent / ".env"
            if env.is_file():
                for raw in env.read_text().splitlines():
                    line = raw.strip()
                    if not line or line.startswith("#") or "=" not in line:
                        continue
                    key, _, value = line.partition("=")
                    os.environ.setdefault(key.strip(), value.strip())
                return


def request(method: str, path: str, params=None, body=None, raw=False) -> int:
    endpoint = os.environ.get("OPENPULSE_ENDPOINT")
    auth = os.environ.get("OPENPULSE_AUTH")
    if not endpoint or not auth or "/" not in auth:
        print("error: OPENPULSE_ENDPOINT and OPENPULSE_AUTH (user/password) must be set", file=sys.stderr)
        return 2
    user, _, password = auth.partition("/")
    url = f"{endpoint.rstrip('/')}{EXTRACTOR_BASE}{path}"
    if params:
        clean = {k: v for k, v in params.items() if v is not None}
        if clean:
            url = f"{url}?{urllib.parse.urlencode(clean)}"
    token = base64.b64encode(f"{user}:{password}".encode()).decode()
    headers = {"Authorization": f"Basic {token}", "Accept": "application/json"}
    data = None
    if body is not None:
        data = json.dumps(body).encode()
        headers["Content-Type"] = "application/json"
    req = urllib.request.Request(url, data=data, headers=headers, method=method)
    try:
        with urllib.request.urlopen(req, timeout=90) as resp:
            out = resp.read()
    except urllib.error.HTTPError as e:
        print(f"http {e.code}: {e.read().decode(errors='replace')}", file=sys.stderr)
        return 1
    except urllib.error.URLError as e:
        print(f"network error: {e.reason}", file=sys.stderr)
        return 1
    except (TimeoutError, socket.timeout):
        print("network error: request timed out", file=sys.stderr)
        return 1
    if raw:
        sys.stdout.buffer.write(out)
        if not out.endswith(b"\n"):
            sys.stdout.write("\n")
        return 0
    try:
        json.dump(json.loads(out), sys.stdout, indent=2, default=str)
        sys.stdout.write("\n")
    except json.JSONDecodeError:
        sys.stdout.buffer.write(out)
        sys.stdout.write("\n")
    return 0


def main() -> int:
    p = argparse.ArgumentParser(description=__doc__.splitlines()[0])
    p.add_argument("command", help="search|stats|manifest|freshness|get")
    p.add_argument("a", nargs="?", help="provider or path")
    p.add_argument("query", nargs="*", help="search query (free text)")
    p.add_argument("--top-k", type=int, dest="top_k")
    p.add_argument("--candidate-k", type=int, dest="candidate_k")
    p.add_argument("--target")
    p.add_argument("--filter", help="metadata filter as JSON object")
    p.add_argument("--sources", action="store_true", help="manifest: include source detail")
    p.add_argument("--param", action="append", help="extra query param key=value (repeatable)")
    p.add_argument("--raw", action="store_true")
    args = p.parse_args()

    load_dotenv()
    extra = {}
    for pair in args.param or []:
        k, _, v = pair.partition("=")
        extra[k.strip()] = v.strip()

    c = args.command
    if c == "search":
        q = " ".join(args.query).strip()
        if not args.a or not q:
            print("error: usage: search <provider> <query>", file=sys.stderr)
            return 2
        body = {"query": q, "top_k": args.top_k, "candidate_k": args.candidate_k, "target": args.target}
        if args.filter:
            body["filter_payload"] = json.loads(args.filter)
        body = {k: v for k, v in body.items() if v is not None}
        return request("POST", f"/v2/indices/{args.a}/search", extra, body=body, raw=args.raw)
    if c == "stats":
        return request("GET", f"/v2/indices/{args.a}/stats", extra, raw=args.raw)
    if c == "manifest":
        return request("GET", "/v2/manifest", {"sources": str(args.sources).lower() if args.sources else None, **extra}, raw=args.raw)
    if c == "freshness":
        return request("GET", "/v2/indices/freshness", extra, raw=args.raw)
    if c == "get":
        path = args.a if args.a.startswith("/") else f"/{args.a}"
        return request("GET", path, extra, raw=args.raw)
    print(f"error: unknown command '{c}'", file=sys.stderr)
    return 2


if __name__ == "__main__":
    sys.exit(main())
